File: data/src/main/java/com/balv/imdb/data/local/dao/MovieDao.java
package com.balv.imdb.data.local.dao;

import androidx.lifecycle.LiveData;
import androidx.paging.DataSource;
import androidx.paging.PagingLiveData;
import androidx.paging.PagingSource;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.balv.imdb.data.model.MovieEntity;

import java.util.List;

import io.reactivex.Observable;

@Dao
public abstract class MovieDao {
    @Query("SELECT * FROM movies")
    public abstract Observable<List<MovieEntity>> getMainMovieList();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertAll(List<MovieEntity> movieEntities);

    @Update
    public abstract void updateMovies(MovieEntity... movieEntities);

    @Query("SELECT * FROM movies WHERE id = :id")
    public abstract Observable<MovieEntity> setMovieDetailSource(String id);

    @Delete
    public abstract void deleteItem(MovieEntity... ids);

    @Query("SELECT COUNT(*) FROM movies")
    public abstract int getCount();

    @Query("DELETE FROM movies")
    public abstract void clearAll();

    @Query("SELECT * FROM movies")
    public abstract PagingSource<Integer, MovieEntity> pagingSource();

}

File: data/src/main/java/com/balv/imdb/data/local/dao/MovieDao.kt
package com.balv.imdb.data.local.dao

import androidx.paging.PagingSource
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import androidx.room.Upsert
import com.balv.imdb.data.model.MovieEntity
import kotlinx.coroutines.flow.Flow

@Dao
abstract class MovieDao {
    @Query("SELECT * FROM movies")
    abstract fun getMainMovieList(): Flow<List<MovieEntity>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertAll(movieEntities: List<MovieEntity>)

    @Upsert
    abstract fun updateMovies(vararg movieEntities: MovieEntity)

    @Query("SELECT * FROM movies WHERE id = :id")
    abstract fun getMovieDetailLocal(id: Int): Flow<MovieEntity?>

    @Delete
    abstract fun deleteItem(vararg ids: MovieEntity?)

    @Query("SELECT COUNT(*) FROM movies")
    abstract suspend fun count(): Int

    @Query("DELETE FROM movies")
    abstract fun clearAll()

    @Query("SELECT * FROM movies")
    abstract fun pagingSource(): PagingSource<Int, MovieEntity>
}
